### PR TITLE
fix: ackHandler order responses under edge cases

### LINKED ISSUE #638: Bug: Under certain `edge` conditions Acks can be sent out of order.
Due to race conditions, is possible for the AckHandler to send Acks out of order.

Cases:

1. When streaming really really fast blocks into the BN (latency = 0)
2. When starting with block higher right after the system has started, meaning the first ack to attempt being sent is 2, and then 1 arrives. (this also breaks the order)

### PR DESCRIPTION
**Pull request by AlfredoG87**
## Reviewer Notes

This PR adds test that helped replicate edge cases when acks were sent out of order.
either due to the workaround to determine the first valid block (where to start sending acks) or
due to random delays occur in either `persistence` or `verification` threads.

Also this PR includes the fixes so those tests always pass instead of always failing or sometimes failing.

Fixes #638

**Before:**
image

**After:** Under the same conditions, same result in multiple passes.
image

### CODE REVIEW

File: server/src/main/java/com/hedera/block/server/ack/AckHandlerImpl.java [hunk 1]
```diff
@@ -111,8 +111,8 @@ public void blockVerificationFailed(long blockNumber) {
     private void attemptAcks() {
         // Temporarily if lastAcknowledgedBlockNumber is -1, we get the first block in the map
         if (lastAcknowledgedBlockNumber == -1) {
-            lastAcknowledgedBlockNumber =
-                    blockInfo.keySet().stream().min(Long::compareTo).orElse(1L) - 1;
+            // todo(147): once we have a way to get the last acknowledged block from the store we should use that
+            lastAcknowledgedBlockNumber = 0;
```
Comment (ata-nas) [inline server/src/main/java/com/hedera/block/server/ack/AckHandlerImpl.java:115]: As noted, as a very temporary solution, here is started from 0, but needs to change once we have the appropriate logic, otherwise the stream cannot start from another block number.

File: server/src/main/java/com/hedera/block/server/ack/AckHandlerImpl.java [hunk 2]
```diff
@@ -137,19 +137,24 @@ private void attemptAcks() {
...
                 blockInfo.remove(nextBlock);
 
-                // Update metrics
+                // Update metrics and logging
                 metricsService.get(BlockNodeMetricTypes.Counter.AckedBlocked).increment();
-            }
+                LOGGER.log(System.Logger.Level.DEBUG, "ACKed block " + nextBlock);
 
+                // Update last acknowledged
+                lastAcknowledgedBlockNumber = nextBlock;
```
Comment (ata-nas) [inline server/src/main/java/com/hedera/block/server/ack/AckHandlerImpl.java:151]: As we noted, the update to the volatile field is the one that will allow other threads to get inside the if check, so this being done last is important.

Comment (AlfredoG87) [reply server/src/main/java/com/hedera/block/server/ack/AckHandlerImpl.java:151]: Correct, this was left to the last step indeed

**Review by mattp-swirldslabs — APPROVED** [review]
It looks like the discussions were resolved. Looks good Fredy

File: server/src/main/java/com/hedera/block/server/ack/AckHandlerImpl.java [hunk 3]
```diff
@@ -111,8 +111,8 @@ public void blockVerificationFailed(long blockNumber) {
     private void attemptAcks() {
         // Temporarily if lastAcknowledgedBlockNumber is -1, we get the first block in the map
         if (lastAcknowledgedBlockNumber == -1) {
-            lastAcknowledgedBlockNumber =
-                    blockInfo.keySet().stream().min(Long::compareTo).orElse(1L) - 1;
+            // todo(147): once we have a way to get the last acknowledged block from the store we should use that
```
Comment (Nana-EC) [inline server/src/main/java/com/hedera/block/server/ack/AckHandlerImpl.java:114]: nit: please link to an actual issue so we have it on the board.
I'm not a fan of to-do that can get lost in code.

**Review by Nana-EC — COMMENTED** [review]
LG
Rebase to resolve conflict and todo comment

File: server/src/main/java/com/hedera/block/server/ack/AckHandlerImpl.java (hunk 3 above)
Comment (AlfredoG87) [reply server/src/main/java/com/hedera/block/server/ack/AckHandlerImpl.java:114]: That is what the `todo(147)` means, 147 is the actual issue number, ohhh am missing the `@` it should be `@todo(147)` 😅  will fix during rebase.

I waited to merge this one on Atanas request, to avoid conflict resolution on his side, since his ticket was way more complicated and large than this one, I yielded in favor of him, but now I can finish it :)

**Review by mattp-swirldslabs — APPROVED** [review]

**Review by Nana-EC — APPROVED** [review]
